### PR TITLE
TST: Add benchmarks

### PR DESCRIPTION
**Pull request by adriangb**
Adds asv based benchmarks. Example output from `poe bench`:

```
Poe => asv run master~1..HEAD && asv compare master HEAD
· Creating environments
· Discovering benchmarks
·· Uninstalling from virtualenv-py3.8
·· Installing 1b5fd748 <benchmarks> into virtualenv-py3.8..
· Running 12 total benchmarks (2 commits * 1 environments * 6 benchmarks)
[  0.00%] · For python-betterproto commit 1b5fd748 <benchmarks>:
[  0.00%] ·· Benchmarking virtualenv-py3.8
[  4.17%] ··· Running (benchmarks.BenchMessage.time_attribute_access--).....
[ 29.17%] ··· benchmarks.BenchMessage.time_attribute_access                                                                                                                              1.09±0.01μs
[ 33.33%] ··· benchmarks.BenchMessage.time_attribute_setting                                                                                                                              5.62±0.2μs
[ 37.50%] ··· benchmarks.BenchMessage.time_attribute_setting_init                                                                                                                         9.48±0.2μs
[ 41.67%] ··· benchmarks.BenchMessage.time_instantiation                                                                                                                                  9.36±0.2μs
[ 45.83%] ··· benchmarks.BenchMessage.time_overhead                                                                                                                                       32.0±0.3ms
[ 50.00%] ··· benchmarks.MemSuite.mem_instance                                                                                                                                                   408
[ 50.00%] · For python-betterproto commit 0028cc38 <master>:
[ 50.00%] ·· Building for virtualenv-py3.8..
[ 50.00%] ·· Benchmarking virtualenv-py3.8
[ 54.17%] ··· Running (benchmarks.BenchMessage.time_attribute_access--).....
[ 79.17%] ··· benchmarks.BenchMessage.time_attribute_access                                                                                                                              1.09±0.01μs
[ 83.33%] ··· benchmarks.BenchMessage.time_attribute_setting                                                                                                                             5.46±0.06μs
[ 87.50%] ··· benchmarks.BenchMessage.time_attribute_setting_init                                                                                                                        9.07±0.01μs
[ 91.67%] ··· benchmarks.BenchMessage.time_instantiation                                                                                                                                  9.31±0.1μs
[ 95.83%] ··· benchmarks.BenchMessage.time_overhead                                                                                                                                       32.4±0.3ms
[100.00%] ··· benchmarks.MemSuite.mem_instance                                                                                                                                                   408

All benchmarks:

       before           after         ratio
     [0028cc38]       [1b5fd748]
     <master>         <benchmarks>
      1.09±0.01μs      1.09±0.01μs     1.01  benchmarks.BenchMessage.time_attribute_access
      5.46±0.06μs       5.62±0.2μs     1.03  benchmarks.BenchMessage.time_attribute_setting
      9.07±0.01μs       9.48±0.2μs     1.04  benchmarks.BenchMessage.time_attribute_setting_init
       9.31±0.1μs       9.36±0.2μs     1.01  benchmarks.BenchMessage.time_instantiation
       32.4±0.3ms       32.0±0.3ms     0.99  benchmarks.BenchMessage.time_overhead
              408              408     1.00  benchmarks.MemSuite.mem_instance
```

This should be useful for performance monitoring, ex: #144

### CODE REVIEW

**Comment by cetanu** [comment]
Looks pretty cool, probably just need to add `asv` to poetry dependencies somewhere, 

and does this need to be included in the build process? Should the output be parsed and validated? I think I will wait for opinions on that

**Comment by adriangb** [comment]
Oh yeah forgot to add the dependency. It should be optional.

The way I've used benchmarks previously, they aren't run as part of CI, although they could be. Largely depends on how long they take to run. A decent option is for whoever is approving the PR to ask for benchmarks _if_ they suspect performance may be affected.

File: pyproject.toml [hunk 1]
```diff
@@ -12,7 +12,7 @@ packages = [
 ]
 
 [tool.poetry.dependencies]
-python = "^3.6"
+python = "^3"
```
Comment (nat-n) [inline pyproject.toml:15]: What's this about? We don't support python < 3.6.

File: benchmarks/benchmarks.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+@dataclass
+class TestMessage(betterproto.Message):
+    foo: int = betterproto.uint32_field(0)
+    bar: str = betterproto.string_field(1)
+    baz: float = betterproto.float_field(2)
+
+class BenchMessage:
+    """
+    An example benchmark that times the performance of various kinds
```
Comment (nat-n) [inline benchmarks/benchmarks.py:13]: 👀

File: benchmarks/benchmarks.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    def setup(self):
+        self.cls = TestMessage
+        self.instance = TestMessage()
+
+    def time_overhead(self):
+        """Overhead in class definition.
+        """
+        for _ in range(100):
+            @dataclass
+            class Message(betterproto.Message):
```
Comment (nat-n) [inline benchmarks/benchmarks.py:25]: maybe nicer to factor this out into a factory function so the class def is only in the file once?

Maybe that would also allow parameterizing the benchmark with different size classes?

File: benchmarks/benchmarks.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        self.cls()
+
+    def time_attribute_access(self):
+        """Time to access an attribute
+        """
+        self.instance.foo
+        self.instance.bar
+        self.instance.baz
+    
+    def time_attribute_setting_init(self):
```
Comment (nat-n) [inline benchmarks/benchmarks.py:42]:
```suggestion
    def time_init_with_values(self):
```

File: benchmarks/benchmarks.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        """Time to set an attribute
+        """
+        self.cls(0, "test", 0.0)
+
+    def time_attribute_setting(self):
+        """Time to set an attribute
```
Comment (nat-n) [inline benchmarks/benchmarks.py:48]:
```suggestion
        """Time to set attributes
```

**Comment by nat-n** [comment]
Nice, this should help guide optimisation work. 

asv should be a dev dependency, and I don't like the look of changing the python version dependency, is this intentional?

File: benchmarks/benchmarks.py (hunk 2 above)
Comment (adriangb) [reply benchmarks/benchmarks.py:13]: Yeah great description huh

File: benchmarks/benchmarks.py (hunk 3 above)
Comment (adriangb) [reply benchmarks/benchmarks.py:25]: Possibly, but in my experience when you try to make "fancy" things happen in benchmarks, your "fancy thing" may inadvertently affect your benchmark results.

File: pyproject.toml (hunk 1 above)
Comment (adriangb) [reply pyproject.toml:15]: Not related to this PR. I'll revert it.

**Comment by adriangb** [comment]
> asv should be a dev dependency, and I don't like the look of changing the python version dependency, is this intentional?

Added asv. That was not an intentional change, I reverted it.

**Comment by adriangb** [comment]
Is it normal for `project.lock` to change for a PR? Or should I not have committed that? I'm a relatively new Poetry user...

**Comment by nat-n** [comment]
> Is it normal for `project.lock` to change for a PR? Or should I not have committed that? I'm a relatively new Poetry user...

If you've added, removed or updated dependencies then yes the lockfile should also be updated to reflect this.

**Review by nat-n — APPROVED** [review]

**Pull request merged**